units/src/timestamp.rs

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! A UNIX timestamp used as the Bitcoin block time.
+//!
+//! Also known as Epoch Time - January 1, 1970.
+//!
+//! This differs from other UNIX timestamps in that we only use non-negative values. The Epoch
+//! pre-dates Bitcoin so timestamps before this are not useful for block timestamps.
+
+use core::ops;
+
+#[cfg(feature = "arbitrary")]
+use arbitrary::{Arbitrary, Unstructured};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// A Bitcoin block timestamp.
+///
+/// > Each block contains a Unix time timestamp. In addition to serving as a source of variation for
+/// > the block hash, they also make it more difficult for an adversary to manipulate the block chain.
+/// >
+/// > A timestamp is accepted as valid if it is greater than the median timestamp of previous 11
+/// > blocks, and less than the network-adjusted time + 2 hours. "Network-adjusted time" is the
+/// > median of the timestamps returned by all nodes connected to you. As a result block timestamps
+/// > are not exactly accurate, and they do not need to be. Block times are accurate only to within
+/// > an hour or two.
+///
+/// ref: <https://en.bitcoin.it/wiki/Block_timestamp>
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Timestamp(u32);
+
+impl Timestamp {
+    /// Constructs a new [`Timestamp`] from an unsigned 32 bit integer value.
+    // Because From<u32> is not const.
+    #[inline]
+    pub const fn from_u32(t: u32) -> Self { Timestamp(t) }
+
+    /// Returns the inner `u32` value.
+    // Because type inference doesn't always work using `Into`.
+    #[inline]
+    pub const fn to_u32(self) -> u32 { self.0 }
+}
+
+impl From<u32> for Timestamp {
+    fn from(t: u32) -> Self { Self::from_u32(t) }
+}
+
+impl From<Timestamp> for u32 {
+    fn from(t: Timestamp) -> Self { t.to_u32() }
+}
+
+impl ops::Sub for Timestamp {
+    type Output = u32;
+
+    fn sub(self, rhs: Timestamp) -> u32 { self.to_u32() - rhs.to_u32() }
+}
+
```

Comment:
I think we could have conversion to `std::time::SystemTime` but should we? I can't think of a reason not to but I may be missing something, I'm not very bright today.

Also we might have the reverse conversion as `TryFrom`, however I would even consider an *inherent method* (not `From` trait) that rounds and panics if the time is above `u32::MAX` for a very simple reason that it implies either the clock being horribly misconfigured or the software not being adapted to the timestamp hard fork and as such it's useless garbage from that point.

Reply:
I think we should hold off on converting to `SystemTime`. You can do it indirectly by going through `u32`, and my *guess* is that anybody actually dealing with timestamps has a dep on `chrono` or `time` or something and doesn't really care about the half-baked std time stuff. It's easy to add later if it turns out though that people are using it.

(We should continue to support/use `std::time::Duration`, which is the one part of std time that's kinda good and widely used.)

Reply:
> half-baked std time stuff

Well, the bitcoin stuff is equally half-baked. I'm fine with holding off. I'd be also fine to add an optional conversion to a popular stable time-related library if there's demand for it.

Also, I suspect that people might be using `SystemTime` to check block validity. (But there's also the median thing which I'm not sure if it includes the node doing the computation.)

Reply:
Agreed on all counts. Let's hold off til we get some feedback about what people want here.